### PR TITLE
fixes #22866; fixes #19998; ensure destruction for Object construction with custom destructors

### LINKED ISSUE #19998: destructor may not be called for temporary objects
### Example
```nim
type A = object

proc `=destroy`(a: var A) =
  echo "destructor"
 
proc take(a: A) = discard

take A()
echo "scope end"
```

### Current Output
```
scope end
```

### Expected Output
```
scope end
destructor
```
or, like in c++
```
destructor
scope end
```

### Additional Information
current behavior is same with all optimizations and memory management strategies
```
$ nim -v
Nim Compiler Version 1.7.1 [Linux: amd64]
Compiled at 2022-07-11
Copyright (c) 2006-2022 by Andreas Rumpf

git hash: a90763ebd762a4862e2af1dab7d4429f4e984a8a
active boot switches: -d:release
```

### LINKED ISSUE #22866: Destructor not called for temp objects passed as arguments
### Description

```nim
type
  SmartPtr = object
    o: int

proc newId(): int =
  var g {.global.} = 0
  inc g
  echo "new ", g
  g

proc `=destroy`*(o: var SmartPtr) =
  if o.o != 0:
    echo "destroy ", o.o
    o.o = 0

proc `=copy`*(a: var SmartPtr, b: SmartPtr) =
  if a.o != 0:
    echo "delete ", a.o
  if b.o == 0:
    a.o = 0
  else:
    a.o = newId()

proc foo(s: SmartPtr) =
  discard

proc test() =
  var a = SmartPtr(o: newId())
  foo(SmartPtr(o: newId()))

test()
```

### Nim Version

```
Nim Compiler Version 2.1.1 [Linux: amd64]
Compiled at 2023-10-22
Copyright (c) 2006-2023 by Andreas Rumpf

git hash: c13c48500b527cb483480702390b1f230ac896a4
active boot switches: -d:release
```

### Current Output

```text
new 1
new 2
destroy 1
```

### Expected Output

Should contain `destroy 2` somewhere.

### Possible Solution

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by ringabout**
fixes #22866;
fixes #19998

### CODE REVIEW

File: compiler/injectdestructors.nim [hunk 1]
```diff
@@ -857,7 +857,7 @@ proc p(n: PNode; c: var Con; s: var Scope; mode: ProcessMode; tmpFlags = {sfSing
             result[i][1] = p(n[i][1], c, s, m)
         else:
           result[i] = p(n[i], c, s, m)
-      if mode == normal and isRefConstr:
+      if mode == normal and hasDestructor(c, n.typ):
```
Comment (ringabout) [inline compiler/injectdestructors.nim]:
```suggestion
      if mode == normal and hasDestructor(c, t):
```

File: compiler/injectdestructors.nim [hunk 2]
```diff
@@ -857,7 +857,7 @@ proc p(n: PNode; c: var Con; s: var Scope; mode: ProcessMode; tmpFlags = {sfSing
             result[i][1] = p(n[i][1], c, s, m)
         else:
           result[i] = p(n[i], c, s, m)
-      if mode == normal and isRefConstr:
+      if mode == normal and hasDestructor(c, t):
```
Comment (ringabout) [inline compiler/injectdestructors.nim]:
```suggestion
      if mode == normal and (isRefConstr or (hasDestructor(c, t) and getAttachedOp(c.graph, t, attachedDestructor) != nil and
        sfOverridden in getAttachedOp(c.graph, t, attachedDestructor).flags)):
```

File: compiler/injectdestructors.nim [hunk 3]
```diff
@@ -857,7 +857,8 @@ proc p(n: PNode; c: var Con; s: var Scope; mode: ProcessMode; tmpFlags = {sfSing
             result[i][1] = p(n[i][1], c, s, m)
         else:
           result[i] = p(n[i], c, s, m)
-      if mode == normal and isRefConstr:
+      if mode == normal and (isRefConstr or (hasDestructor(c, t) and getAttachedOp(c.graph, t, attachedDestructor) != nil and
+        sfOverridden in getAttachedOp(c.graph, t, attachedDestructor).flags)):
```
Comment (ringabout) [inline compiler/injectdestructors.nim]:
```suggestion
      if mode == normal and (isRefConstr or (hasDestructor(c, t) and
        getAttachedOp(c.graph, t, attachedDestructor) != nil and
        sfOverridden in getAttachedOp(c.graph, t, attachedDestructor).flags)):
```

**Comment by Araq** [comment]
Please backport @narimiran

**Pull request merged**